### PR TITLE
Add WorkProcessor#transformProcessor and dispose unused resources

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessor.java [hunk 1]
```diff
@@ -86,6 +86,11 @@
         return WorkProcessorUtils.transform(this, transformation);
     }
 
+    default <R> WorkProcessor<R> transformProcessor(Function<WorkProcessor<T>, WorkProcessor<R>> transformation)
```
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/operator/WorkProcessor.java]: Java gets it confused with `transform(Transformation<T, R> transformation)` when using as lambda

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java [hunk 2]
```diff
@@ -173,7 +173,7 @@ else if (processor.isBlocked()) {
     {
         requireNonNull(processor, "processor is null");
         requireNonNull(transformation, "transformation is null");
-        return flatten(processor.transform(transformation));
+        return processor.transform(transformation).transformProcessor(WorkProcessorUtils::flatten);
```
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java]: > what's the gain?

To be able to use streamlined transformation of the whole `processor` as here.
One could imagine having transformations that provide few steps, e.g:
```
WorkProcessor<WIndowPartition> orderedPagesToWindowPartition(WorkProcessor<Page> pages) {
  return pages
    .transform(new PagesIndexFromOrderedPages(...)).
    .map(pagesIndex -> ...);
} 

processor.transformProcessor(Class::orderedPagesToWindowPartition)
```

> eg i would remove mergeSorted from the interface, as it's likely not as useful as other methods for different users of the interface

sorting methods are part of https://google.github.io/guava/releases/21.0/api/docs/com/google/common/collect/Iterables.html and https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html
which I modeled from

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessor.java [hunk 3]
```diff
@@ -77,14 +77,18 @@
...
+     */
     default <R> WorkProcessor<R> transform(Transformation<T, R> transformation)
     {
         return WorkProcessorUtils.transform(this, transformation);
     }
 
     /**
      * Converts {@link WorkProcessor} into an {@link Iterator}. The iterator will throw {@link IllegalStateException} when underlying {@link WorkProcessor}
-     * yields or becomes blocked.
+     * yields or becomes blocked. {@link WorkProcessor} instance will be dereferenced immediately after iterator is finished.
```
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/operator/WorkProcessor.java]: `" .. after iterator is exhausted"` changed

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java [hunk 4]
```diff
@@ -236,51 +246,64 @@ else if (processor.isBlocked()) {
...
-        requireNonNull(process, "process is null");
-        return new WorkProcessor<T>()
+        return new ProcessWorkProcessor<>(process);
+    }
+
+    private static class ProcessWorkProcessor<T>
+            implements WorkProcessor<T>
+    {
+        WorkProcessor.Process<T> process;
+        ProcessorState<T> state;
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java]: no need for Optional, it wouldn't make the code more readable. What about NEEDS_MORE_DATA as the initial state? (i don't insist)

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java [hunk 5]
```diff
@@ -59,28 +59,38 @@ protected T computeNext()
...
     {
-        requireNonNull(processor, "processor is null");
-        return new AbstractIterator<Optional<T>>()
+        return new YieldingIterator<>(processor);
+    }
+
+    private static class YieldingIterator<T>
+            extends AbstractIterator<Optional<T>>
+    {
+        WorkProcessor<T> processor;
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java]: bump - private

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java [hunk 6]
```diff
@@ -236,51 +246,64 @@ else if (processor.isBlocked()) {
...
     {
-        requireNonNull(process, "process is null");
-        return new WorkProcessor<T>()
+        return new ProcessWorkProcessor<>(process);
+    }
+
+    private static class ProcessWorkProcessor<T>
+            implements WorkProcessor<T>
+    {
+        WorkProcessor.Process<T> process;
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java]: bump - private

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java]: make it privae

**Pull request merged**